plugins/modules/cargo.py

```diff
@@ -163,19 +178,53 @@ def install(self, packages=None):
         if self.version:
             cmd.append("--version")
             cmd.append(self.version)
+        if self.directory:
+            cmd.append("--path")
+            cmd.append(self.directory)
         return self._exec(cmd)
 
     def is_outdated(self, name):
         installed_version = self.get_installed().get(name)
+        latest_version = (
+            self.get_latest_published_version(name)
+            if not self.directory
+            else self.get_source_directory_version(name)
+        )
+        return installed_version != latest_version
 
+    def get_latest_published_version(self, name):
         cmd = ["search", name, "--limit", "1"]
         data, dummy = self._exec(cmd, True, False, False)
 
         match = re.search(r'"(.+)"', data)
-        if match:
-            latest_version = match.group(1)
-
-        return installed_version != latest_version
+        if not match:
+            self.module.fail_json(
+                msg="No published version for package %s found" % name
+            )
```

Comment:
Hmm, this is a breaking change, since the module now fails if no published version for a package is found. It didn't do that before.

(How about emitting a warning in that case instead? You can use `self.module.warn(msg)` for that.)

Reply:
Unless I'm missing something: currently, if there is no match, there's going to be a `UnboundLocalError`, as `latest_version` only gets a value binded to it in the match case, causing the subsequent inequality check to error if it doesn't get set.

I guess it's very unlikely that a non-match is going to happen (would require all published versions of a package to be pulled?).

I'm happy to warn and return something, or to just revert back and not touch this. Let me know what you'd like.

Reply:
Ah, I see, I didn't notice that `latest_version` didn't get any value assigned outside the `if`... Calling `fail_json` then is definitely better!